File: ruby/007/007.rb
def sieve_of_atkin(limit)
  primes = []

  sieve = Array.new(limit + 1) { false }

  square_root_of_limit = Math.sqrt(limit).to_i

  (1..square_root_of_limit).each do |x|
    (1..square_root_of_limit).each do |y|
      n = (4 * x * x) + (y * y)

      sieve[n] = !sieve[n] if n <= limit && (n % 12 == 1 || n % 12 == 5)

      n = (3 * x * x) + (y * y)

      sieve[n] = !sieve[n] if n <= limit && n % 12 == 7

      n = (3 * x * x) - (y * y)

      sieve[n] = !sieve[n] if x > y && n <= limit && n % 12 == 11
    end
  end

  sieve[2] = true

  sieve[3] = true

  primes << 2

  primes << 3

  n = 5

  while n <= square_root_of_limit
    if sieve[n]
      i = n * n

      while i < limit
        sieve[i] = false

        i += n * n
      end

      primes << n
    end

    n += 2
  end

  while n < limit
    primes << n if sieve[n]

    n += 2
  end

  primes
end

n = 10001

limit = ((n * Math.log(n)) + (n * Math.log(Math.log(n)))).to_i

puts sieve_of_atkin(limit)[n - 1]

File: c/007/007.c
#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include "unsigned_long_long_array.h"

void sieveOfAtkin(UnsignedLongLongArray *primes, unsigned long long limit);

int main(int argc, char *argv[]) {
  unsigned long long n = 10001;

  unsigned long long limit = (unsigned long long)((n * log(n)) + (n * (log(log(n)))));

  UnsignedLongLongArray primes;

  UnsignedLongLongArrayInitialize(&primes);

  sieveOfAtkin(&primes, limit);

  printf("%llu\n", primes.values[n - 1]);

  UnsignedLongLongArrayDeinitialize(&primes);

  return 0;
}

void sieveOfAtkin(UnsignedLongLongArray *primes, unsigned long long limit) {
  bool *sieve = calloc(limit, sizeof(bool));

  unsigned long long squareRootOfLimit = (unsigned long long)sqrt(limit);

  unsigned long long n;

  for (unsigned long long x = 1; x <= squareRootOfLimit; x++) {
    for (unsigned long long y = 1; y <= squareRootOfLimit; y++) {
      n = (4 * x * x) + (y * y);

      if (n <= limit && (n % 12 == 1 || n % 12 == 5)) {
        sieve[n] = !sieve[n];
      }

      n = (3 * x * x) + (y * y);

      if (n <= limit && n % 12 == 7) {
        sieve[n] = !sieve[n];
      }

      n = (3 * x * x) - (y * y);

      if (x > y && n <= limit && n % 12 == 11) {
        sieve[n] = !sieve[n];
      }
    }
  }

  sieve[2] = true;

  sieve[3] = true;

  UnsignedLongLongArrayPush(primes, 2);

  UnsignedLongLongArrayPush(primes, 3);

  n = 5;

  while (n <= squareRootOfLimit) {
    if (sieve[n]) {
      unsigned long long i = n * n;

      while (i < limit) {
        sieve[i] = false;

        i += n * n;
      }

      UnsignedLongLongArrayPush(primes, n);
    }

    n += 2;
  }

  while (n < limit) {
    if (sieve[n]) {
      UnsignedLongLongArrayPush(primes, n);
    }

    n += 2;
  }

  free(sieve);
}
